walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,174 @@
+    attendees.removeAll(Collections.singleton(""));
+    optionalAttendees.removeAll(Collections.singleton(""));
+
+    long duration = request.getDuration();
+
+    // Populate {@code attendeesEventTimes} and {@code optionalAttendeesEventTimes}.
+    getAttendeesEventTimes(events, attendees, optionalAttendees);
+
+    Collection<TimeRange> possibleTimes = new ArrayList<>();
+    if (attendees.isEmpty()) {
+      // If there is no mandatory attendee, only consider availabilities of optional attendees.
+
+      // {@code eventTimes} is the list of current events' time ranges that need to be avoided when scheduling.
+      List<TimeRange> eventTimes = new ArrayList<>();
+      for (List<TimeRange> timeList : optionalAttendeesEventTimes.values()) {
+        eventTimes.addAll(timeList);
+      }
+      Collections.sort(eventTimes, TimeRange.ORDER_BY_START);
+      possibleTimes = findTimeRangeGaps(eventTimes, duration);
+    } else {
+      // If there are mandatory attendees, consider optional attendees who can possibly attend and ignore those who can't. 
+
+      List<TimeRange> eventTimes = new ArrayList<>(attendeesEventTimes);
+
+      // Find the time slot(s) that maximize the number of optional attendees who can attend.
+      // Count down from the maximum number of optional attendees and find all combinations of that number of attendees.
+      // Once we find a combination where scheduling is possible, we've found the maximal group of attendees.
+      for (int num = optionalAttendees.size(); num > 0; num--) {
+        String[] arr = optionalAttendees.toArray(new String[optionalAttendees.size()]);
+        getAllCombinations(arr, arr.length, num);
+
+        for (List<String> optionalAttendeesChosen : optionalAttendeesCombs) {
+          for (String attendee : optionalAttendeesChosen) {
+            eventTimes.addAll(optionalAttendeesEventTimes.get(attendee));
+          }
+          Collections.sort(eventTimes, TimeRange.ORDER_BY_START);
+          possibleTimes = findTimeRangeGaps(eventTimes, duration);
+          if (!possibleTimes.isEmpty()) {
+            // We've found a combination where scheduling is possible, so those are the desired time slot(s).
+            return possibleTimes;
+          }
+          // Re-initialize unavailable time ranges with those of the mandatory attendees.
+          eventTimes = new ArrayList<>(attendeesEventTimes);
+        }
+      }
+      // Calculate time ranges when there are no optional attendees, or none of the optional attendees can fit the schedule.
+      possibleTimes = findTimeRangeGaps(eventTimes, duration);
+    }
+    return possibleTimes;
+  }
+
+  /**
+   * Recursive helper function used to find all combinations of size r of the input array arr[].
+   * chosen[]: temporary array to store current combination.
+   * start & end: staring and ending indexes in arr[].
+   * index: current index in data[].
```

Comment:
`@param` Java docs.

Also, each param needs to be mentioned, on its own line.

Reply:
got it